### ISSUE TITLE
Application Namespace Not Updated for Flex Desktop Build & Run

### ISSUE THREAD

**Issue by JoelProminic** [Moonshine-IDE/Moonshine-IDE#869]
@feather812002 reported an error like this on macOS 10.13 and 10.14 with recent nightly builds.  I can confirm part of it on macOS 10.15.7:
1. Create a new Flex Desktop project.  Select the Harman AIR (33.1) SDK as the default SDK
1. Open the `src/%project%-app.xml` file.  The namespace is set to 32.0  (`<application xmlns="http://ns.adobe.com/air/application/32.0">`).  Is this hard-coded in the template?
1. Update the namespace to 33.1
1. Project > Build & Run.  BUG:  in some cases, the application fails to start here
1. Open `bin-debug/%project%-app.xml`.  BUG:  the namespace is 32.0 rather than 33.1
1. Update the namespace to 33.1
1. Run the application manually.  Open `bin-debug` in a Terminal and run `~/Downloads/MoonshineSDKs/Flex_SDK/Flex_4.16.1_AIR_33.1/bin/adl %project%-app.xml`.  This runs properly

When @feather812002 tries to build Moonshine, it shows the Harman splash screen on Build & Run, but the app never opens.  Fixing the namespace allows the application to launch.

How is the namespace updated for the `bin-debug` copy of the descriptor file?

**Comment by rat-moonshine** [comment]
> Open the src/%project%-app.xml file. The namespace is set to 32.0 (<application xmlns="http://ns.adobe.com/air/application/32.0">). Is this hard-coded in the template?

Moonshine generates the namespace to the application descriptor file during a project's creation - based on the default or given SDK's configuration-read value.

For this Moonshine relies on `$sdkPath/frameworks/flex-config.xml` 's `<taget-player>` value. This value also uses during a _Build & Run_ process, to update the `bin-debug` copied `*-app.xml` file's namespace based on the currently running SDK's major-player-version. 

I saw this value is 32.0 in case of downloaded/installed Harman Flex/AIR SDK. I see I did some fixes here but this never been tested - https://github.com/Moonshine-IDE/Moonshine-SDK-Installer/issues/56#issuecomment-862357089.

**Comment by rat-moonshine** [comment]
I did couple of changes which affects following areas:

1. `namespace` generate based on selected SDK during a project's creation - previously, it was always populated by Moonshine's default-SDK value and based on custom project being selected (during a project's creation)
2. `namespace` generation with proper `{major}.{minor}` values - updated to work for Harman and previous Adobe SDKs
3. Proper `namespace` generation in `*-app.xml` file that being copied into `bin-debug` folder during a "Build & Run"
4. Couple of other areas updated to keep the major and minor version of swf using the current SDK

Beside that I have updated the Moonshine SDK Installer to install Harman Flex/AIR SDK with proper `target-player` value in its `flex-config.xml` file. Should be tested with newer installation (of MSDKI).

**Comment by rat-moonshine** [comment]
During our tests with @JoelProminic we found that a mirror-URL to download Apache Flex SDK binary is broken. (A corresponding issue has created to choose between available mirrors here https://github.com/Moonshine-IDE/Moonshine-SDK-Installer/issues/62.) 

For now, I have updated the mirror-URLs to download Apache Flex SDK to its backup site - it looks like the mirror-URL suggests/provides by Apache based on geolocation and mine was a `.in` mirror which might not be very correct for others in different sides of the world, so I chose the backup site (https://downloads.apache.org) for now.

**Comment by JoelProminic** [comment]
I tested this on macOS, and it seems to be working well.

Custom SDK | Default SDK | Populated Namespace
-----------|---------------|----------
Harman | AIR | 33.1
AIR | AIR | 33.1
\- | Harman | 33.1
\- | AIR | 32.0

Windows needs testing still, but this is a good sign.

**Comment by JoelProminic** [comment]
@feather812002 reported that he was still seeing this error with the latest Moonshine build.
![image](https://user-images.githubusercontent.com/24532731/124975332-8963d400-dff3-11eb-834f-dd3e79a1f611.png)

@feather812002, you'll need to delete your ~/Downloads/MoonshineSDKs/Flex_SDK/Flex_4.16.1_AIR_33.1 directory and download it again with a [newer build of MSDKI](https://moonshine-ide.com/downloads/msdki/dev/macos/) to get some required SDK-side changes.

Another difference I noticed is that @feather812002 was using an external JDK 1.8 for Java Development Kit Root Path.   I did a quick test with an external JDK 8 on my normal workstation, and I found that application could still open properly, so I don't think that this is the problem.   However, I recommend installing and configuring JDK 11 using MSDKI anyway, to avoid problems with the language server..  If this download doesn't work for you for some reason, let me know so that we can investigate further.

**Comment by JoelProminic** [comment]
@feather812002, have you tried testing this with a fresh Harman AIR installation?   Please check this, so that we can close the issue if that fixes your problem.s

**Comment by piotrzarzycki21** [comment]
Hey @feather812002 could you please test this ? Download newest Nightly Moonshine SDK Installer, install new Harman Air. Additionally make sure that you have setup JAVA.

**Comment by rat-moonshine** [comment]
As part of #56 changes namespace generation for -app.xml file in bin-debug now retrieves from `airsdk.xml` instead of earlier `flex-config.xml`. Give this a test if you can running Harman and non-Harman SDK project in Moonshine.

**Comment by feather812002** [comment]
@JoelProminic  after I update the `ide/MoonshineGUICore/bin/flash/bin/MoonshineGUICore.swc `
I can pass the error that I report on yesterday. Now, it can build success.

@piotrzarzycki21  after I install the latest Moonshine-SDK-Installer and latest Harman SDK .
Now, it can auto lunch success and no any error.

**Comment by JoelProminic** [comment]
I installed the latest Moonshine builds and reinstalled the Flex SDKs with the latest version of MSDKI.  Here are my results:

OS | Flex SDK | Namespace | Build & Run
---|----------|-----------|------------
macOS | Adobe AIR | 32.0 | Works
macOS | Harman AIR | 33.1 | Works
Windows | Adobe AIR | 32.0 | Works
Windows | Harman AIR | 33.1 | Works

This looks good to me.

**Comment by rat-moonshine** [comment]
I found a rather unexpected problem while testing Prominic Native in Moonshine-IDE. Even after the SWF compiled well, Moonshine failed to launch the application. 

I've found that `<initialWindow><content>` property was holding the default value and not updated with the SWF file reference.

While looking closely I found there was two `<content>` tag in the file - one comment-off and one active. Moonshine was replacing this tag using `regExp` but without a global parameter. Thus, it was replacing only the first `<content>` which was unfortunately the comment-off one. I adjusted the update codes with a global parameter now which should fix this matter. Tested working correctly for Native now.
